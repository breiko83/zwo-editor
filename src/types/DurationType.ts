export type DurationType = "time" | "distance";