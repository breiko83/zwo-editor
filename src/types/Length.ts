export class Duration {
  constructor(public readonly seconds: number) {}
}

export class Distance {
  constructor(public readonly meters: number) {}
}

export type Length = Distance | Duration;
