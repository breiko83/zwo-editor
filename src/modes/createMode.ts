import { LengthType } from "../types/LengthType";
import { RunningTimes } from "../types/RunningTimes";
import { SportType } from "../types/SportType";
import BikeMode from "./BikeMode";
import RunMode from "./RunMode";
import { WorkoutMode } from "./WorkoutMode";

interface CreateModeParams {
  sportType: SportType;
  ftp: number;
  weight: number;
  runningTimes: RunningTimes;
  lengthType: LengthType;
}

// Creates appropriate type of mode for current sport type
export default function createMode({sportType, ftp, weight, runningTimes, lengthType}: CreateModeParams): WorkoutMode {
  switch (sportType) {
    case "bike": return new BikeMode(ftp, weight);
    case "run": return new RunMode(runningTimes, lengthType);
  }
}
