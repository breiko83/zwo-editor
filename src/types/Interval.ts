import { Length } from "./Length";
import { PaceType } from "./PaceType";

export interface SteadyInterval {
  id: string,
  length: Length,
  type: 'steady',
  intensity: number,
  cadence: number,
  pace: PaceType,
}

export interface RampInterval {
  id: string,
  length: Length,
  type: 'ramp',
  startIntensity: number,
  endIntensity: number,
  cadence: number,
  pace: PaceType,
}

export interface FreeInterval {
  id: string,
  length: Length,
  type: 'free',
  cadence: number,
}

export interface RepetitionInterval {
  id: string,
  type: 'repetition',
  onCadence: number,
  offCadence: number,
  onIntensity: number,
  offIntensity: number,
  onLength: Length,
  offLength: Length,
  repeat: number,
  pace: PaceType,
}

export type Interval = SteadyInterval | RampInterval | FreeInterval | RepetitionInterval;
