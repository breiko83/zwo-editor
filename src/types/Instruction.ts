import { v4 as uuidv4 } from "uuid";
import { WorkoutMode } from "../modes/WorkoutMode";
import { Distance, Duration, Length } from "./Length";

export interface Instruction {
  id: string,
  text: string,
  offset: Length,
}

export function createInstruction(fields: Partial<Instruction>, mode: WorkoutMode): Instruction {
  return {
    id: uuidv4(),
    text: "",
    offset: mode.lengthType === "time" ? new Duration(0) : new Distance(0),
    ...fields,
  };
}
