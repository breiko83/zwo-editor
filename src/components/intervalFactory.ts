import { v4 as uuidv4 } from 'uuid'
import { PaceType } from './Editor/PaceSelector';
import { FreeInterval, Interval, RampInterval, RepetitionInterval, SteadyInterval } from "./Interval";

const defaultDuration = 300;
const defaultPower = 1.0;
const defaultPace = PaceType.oneMile;
const defaultCadence = 0;

export default {
  steady(interval: Partial<SteadyInterval>): SteadyInterval {
    return {
      type: 'steady',
      id: uuidv4(),
      duration: defaultDuration,
      power: defaultPower,
      cadence: defaultCadence,
      pace: defaultPace,
      ...interval,
    };
  },

  ramp(interval: Partial<RampInterval>): RampInterval {
    return {
      type: 'ramp',
      id: uuidv4(),
      duration: defaultDuration,
      startPower: defaultPower / 2,
      endPower: defaultPower,
      cadence: defaultCadence,
      pace: defaultPace,
      ...interval,
    };
  },

  free(interval: Partial<FreeInterval>): FreeInterval {
    return {
      type: 'free',
      id: uuidv4(),
      duration: defaultDuration,
      cadence: defaultCadence,
      ...interval
    };
  },

  repetition(interval: Partial<RepetitionInterval>): RepetitionInterval {
    return {
      type: 'repetition',
      id: uuidv4(),
      cadence: defaultCadence,
      restingCadence: defaultCadence,
      repeat: 3,
      onDuration: 30,
      offDuration: 120,
      onPower: defaultPower,
      offPower: defaultPower / 2,
      pace: defaultPace,
      ...interval,
    };
  },

  // Creates a copy of the interval, with a new ID
  clone(interval: Interval): Interval {
    return { ...interval, id: uuidv4() };
  },
};
