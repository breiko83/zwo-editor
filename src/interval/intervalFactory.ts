import { v4 as uuidv4 } from 'uuid'
import { PaceType } from '../types/PaceType';
import { FreeInterval, Interval, RampInterval, RepetitionInterval, SteadyInterval } from "../types/Interval";
import { Distance, Duration, Length } from '../types/Length';
import { WorkoutMode } from '../modes/WorkoutMode';

const defaultDuration = new Duration(300);
const defaultDistance = new Distance(1000);
const defaultIntensity = 1.0;
const defaultPace = PaceType.oneMile;
const defaultCadence = 0;

const defaultLength = (mode: WorkoutMode): Length =>
  mode.lengthType === "time" ? defaultDuration : defaultDistance;

export default {
  steady(interval: Partial<SteadyInterval>, mode: WorkoutMode): SteadyInterval {
    return {
      type: 'steady',
      id: uuidv4(),
      length: defaultLength(mode),
      intensity: defaultIntensity,
      cadence: defaultCadence,
      pace: defaultPace,
      ...interval,
    };
  },

  ramp(interval: Partial<RampInterval>, mode: WorkoutMode): RampInterval {
    return {
      type: 'ramp',
      id: uuidv4(),
      length: defaultLength(mode),
      startIntensity: defaultIntensity / 2,
      endIntensity: defaultIntensity,
      cadence: defaultCadence,
      pace: defaultPace,
      ...interval,
    };
  },

  free(interval: Partial<FreeInterval>, mode: WorkoutMode): FreeInterval {
    return {
      type: 'free',
      id: uuidv4(),
      length: defaultLength(mode),
      cadence: defaultCadence,
      ...interval
    };
  },

  repetition(interval: Partial<RepetitionInterval>, mode: WorkoutMode): RepetitionInterval {
    return {
      type: 'repetition',
      id: uuidv4(),
      onCadence: defaultCadence,
      offCadence: defaultCadence,
      repeat: 3,
      onLength: mode.lengthType === "time" ? new Duration(30) : new Distance(200),
      offLength: mode.lengthType === "time" ? new Duration(120) : new Distance(200),
      onIntensity: defaultIntensity,
      offIntensity: defaultIntensity / 2,
      pace: defaultPace,
      ...interval,
    };
  },

  // Creates a copy of the interval, with a new ID
  clone(interval: Interval): Interval {
    return { ...interval, id: uuidv4() };
  },
};
