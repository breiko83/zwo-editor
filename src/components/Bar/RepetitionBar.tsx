import React, { useMemo, useState } from 'react'
import SteadyBar from './SteadyBar'
import './RepetitionBar.css'
import { RepetitionInterval, SteadyInterval } from '../../types/Interval'
import { WorkoutMode } from '../../modes/WorkoutMode'
import intervalFactory from '../../interval/intervalFactory'
import { range } from 'ramda'

interface RepetitionBarProps {
  interval: RepetitionInterval;
  mode: WorkoutMode;
  onChange: (interval: RepetitionInterval) => void;
  onClick: (id: string) => void;
  selected: boolean;
}

const RepetitionBar = ({interval, ...props}: RepetitionBarProps) => {
  const [repeat, setRepeat] = useState(interval.repeat)

  const [onLength, setOnLength] = useState(interval.onLength)
  const [offLength, setOffLength] = useState(interval.offLength)

  const subIntervals = useMemo(() => {
    return range(0, repeat).flatMap(() => [
      intervalFactory.steady({
        length: onLength,
        intensity: interval.onIntensity,
        cadence: interval.onCadence,
        pace: interval.pace,
      }, props.mode),
      intervalFactory.steady({
        length: offLength,
        intensity: interval.offIntensity,
        cadence: interval.offCadence,
        pace: interval.pace,
      }, props.mode),
    ]);
    // eslint-disable-next-line
  }, [repeat]);

  function handleOnChange(values: SteadyInterval) {
    const index = subIntervals.findIndex(sub => sub.id === values.id)

    if (index % 2 === 1) {
      setOffLength(values.length)
    }else{
      setOnLength(values.length)
    }

    for (var i = 0; i < subIntervals.length; i++) {
      if (index % 2 === i % 2) {
        subIntervals[i].length = values.length
        subIntervals[i].intensity = values.intensity
        subIntervals[i].cadence = values.cadence
      }
    }

    props.onChange({
      ...interval,
      onCadence: subIntervals[0].cadence,
      offCadence: subIntervals[1].cadence,
      repeat: repeat,
      onLength: subIntervals[0].length,
      offLength: subIntervals[1].length,
      onIntensity: subIntervals[0].intensity,
      offIntensity: subIntervals[1].intensity,
    })
  }

  function handleAddInterval() {
    setRepeat(repeat + 1)
    props.onChange({
      ...interval,
      repeat: interval.repeat + 1,
    })
  }

  function handleRemoveInterval() {
    if (repeat > 1) {
      setRepeat(repeat - 1)
      props.onChange({
        ...interval,
        repeat: interval.repeat - 1,
      })
    }
  }

  const renderBar = (subInterval: SteadyInterval, withLabel: boolean) => (
    <SteadyBar
      key={subInterval.id}
      interval={subInterval}
      mode={props.mode}
      onChange={handleOnChange}
      onClick={() => props.onClick(interval.id)}
      selected={props.selected}
      showLabel={withLabel}
    />
  )

  return (
    <div>
      <div className='buttons'><button onClick={handleAddInterval}>+</button><button onClick={handleRemoveInterval}>-</button></div>
      <div className='intervals'>
        {subIntervals.map((sub, index) => renderBar(sub, index === 0 || index === subIntervals.length - 1))}
      </div>
    </div>
  )
}

export default RepetitionBar
