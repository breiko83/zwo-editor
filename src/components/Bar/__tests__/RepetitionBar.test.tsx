import React from 'react';
import RepetitionBar from '../RepetitionBar';
import renderer from 'react-test-renderer';
import '@testing-library/jest-dom/extend-expect'
import intervalFactory from '../../../interval/intervalFactory';
import createMode from '../../../modes/createMode';
import { Duration } from '../../../types/Length';

const MockReact = React;
let mockId = 0;

jest.mock('../SteadyBar', () => (props: any) =>
  MockReact.createElement("SteadyBar", props));

jest.mock('uuid', () => ({
  v4: () => `mock-id-${++mockId}`,
}));

describe('<RepetitionBar>', () => {
  it('renders using SteadyBar components', () => {
    const mode = createMode({sportType: "bike", ftp: 200, weight: 75, runningTimes: [], lengthType: "time"});
    const interval = intervalFactory.repetition({
      repeat: 3,
      onLength: new Duration(50),
      offLength: new Duration(100),
      onIntensity: 1.0,
      offIntensity: 0.5,
    }, mode);
  
    const component = renderer.create(
      <RepetitionBar
        interval={interval}
        mode={mode}
        selected={false}
        onChange={() => { }}
        onClick={() => { }}
      />
    );

    expect(component).toMatchSnapshot();
  });
});
