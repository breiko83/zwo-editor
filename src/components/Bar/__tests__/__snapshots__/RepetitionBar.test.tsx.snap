// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RepetitionBar> renders using SteadyBar components 1`] = `
<div>
  <div
    className="buttons"
  >
    <button
      onClick={[Function]}
    >
      +
    </button>
    <button
      onClick={[Function]}
    >
      -
    </button>
  </div>
  <div
    className="intervals"
  >
    <SteadyBar
      interval={
        Object {
          "cadence": 0,
          "id": "mock-id-2",
          "intensity": 1,
          "length": Duration {
            "seconds": 50,
          },
          "pace": 0,
          "type": "steady",
        }
      }
      mode={
        BikeMode {
          "ftp": 200,
          "lengthType": "time",
          "sportType": "bike",
          "weight": 75,
        }
      }
      onChange={[Function]}
      onClick={[Function]}
      selected={false}
      showLabel={true}
    />
    <SteadyBar
      interval={
        Object {
          "cadence": 0,
          "id": "mock-id-3",
          "intensity": 0.5,
          "length": Duration {
            "seconds": 100,
          },
          "pace": 0,
          "type": "steady",
        }
      }
      mode={
        BikeMode {
          "ftp": 200,
          "lengthType": "time",
          "sportType": "bike",
          "weight": 75,
        }
      }
      onChange={[Function]}
      onClick={[Function]}
      selected={false}
      showLabel={false}
    />
    <SteadyBar
      interval={
        Object {
          "cadence": 0,
          "id": "mock-id-4",
          "intensity": 1,
          "length": Duration {
            "seconds": 50,
          },
          "pace": 0,
          "type": "steady",
        }
      }
      mode={
        BikeMode {
          "ftp": 200,
          "lengthType": "time",
          "sportType": "bike",
          "weight": 75,
        }
      }
      onChange={[Function]}
      onClick={[Function]}
      selected={false}
      showLabel={false}
    />
    <SteadyBar
      interval={
        Object {
          "cadence": 0,
          "id": "mock-id-5",
          "intensity": 0.5,
          "length": Duration {
            "seconds": 100,
          },
          "pace": 0,
          "type": "steady",
        }
      }
      mode={
        BikeMode {
          "ftp": 200,
          "lengthType": "time",
          "sportType": "bike",
          "weight": 75,
        }
      }
      onChange={[Function]}
      onClick={[Function]}
      selected={false}
      showLabel={false}
    />
    <SteadyBar
      interval={
        Object {
          "cadence": 0,
          "id": "mock-id-6",
          "intensity": 1,
          "length": Duration {
            "seconds": 50,
          },
          "pace": 0,
          "type": "steady",
        }
      }
      mode={
        BikeMode {
          "ftp": 200,
          "lengthType": "time",
          "sportType": "bike",
          "weight": 75,
        }
      }
      onChange={[Function]}
      onClick={[Function]}
      selected={false}
      showLabel={false}
    />
    <SteadyBar
      interval={
        Object {
          "cadence": 0,
          "id": "mock-id-7",
          "intensity": 0.5,
          "length": Duration {
            "seconds": 100,
          },
          "pace": 0,
          "type": "steady",
        }
      }
      mode={
        BikeMode {
          "ftp": 200,
          "lengthType": "time",
          "sportType": "bike",
          "weight": 75,
        }
      }
      onChange={[Function]}
      onClick={[Function]}
      selected={false}
      showLabel={true}
    />
  </div>
</div>
`;
